File: Common/JS/DS/binaryTree/traversal.js
class Node {
  constructor(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

let root = new Node(1);
root.left = new Node(2);
root.right = new Node(3, new Node(4), new Node(5));

function preorder(node) {
  if (!node.left && !node.right) {
    console.log(node.val, ' ');
    return;
  }

  console.log(node.val, ' ');
  if (node.left) preorder(node.left);
  if (node.right) preorder(node.right);
}

function inorder(node) {
  if (!node.left && !node.right) {
    console.log(node.val, ' ');
    return;
  }

  if (node.left) inorder(node.left);
  console.log(node.val, ' ');
  if (node.right) inorder(node.right);
}

function postorder(node) {
  if (!node.left && !node.right) {
    console.log(node.val, ' ');
    return;
  }

  if (node.left) postorder(node.left);
  if (node.right) postorder(node.right);
  console.log(node.val, ' ');
}

function levelorder(root) {
  let queue = [root];
  let values = [];
  while (queue.length > 0) {
    node = queue.shift();
    values.push(node.val);

    if (node.left) queue.push(node.left);
    if (node.right) queue.push(node.right);
  }
  console.log(values.join(', '));
}

console.log('preorder');
preorder(root);

console.log('inorder');
inorder(root);

console.log('postorder');
postorder(root);

console.log('levelorder');
levelorder(root);

File: Common/Ruby/DS/binary_tree/traversal.rb
class Node
  attr_accessor :val, :left, :right
  def initialize(val, left=nil, right=nil)
    @val = val
    @left = left
    @right = right
  end
end

root = Node.new(1)
root.left = Node.new(2)
root.right = Node.new(3, Node.new(4), Node.new(5))

#       6
#     4    8
#   2  5  7  9
# 1  3

# Root -> L -> R
# 642135879
def pre_order(node)
  values = []
  if node.left.nil? && node.right.nil?
    values << node.val
    return values
  end

  values << node.val
  values << pre_order(node.left)
  values << pre_order(node.right)

  values
end

# L -> Root -> R
# 對 BST 做 inorder 就是由小到大依序遍歷。
def in_order(node)
  values = []
  if node.left.nil? && node.right.nil?
    values << node.val
    return values
  end

  values << in_order(node.left)
  values << node.val
  values << in_order(node.right)

  values
end

# L -> R -> Root
def post_order(node)
  values = []
  if node.left.nil? && node.right.nil?
    values << node.val
    return values
  end

  values << post_order(node.left)
  values << post_order(node.right)
  values << node.val

  values
end

def level_order(root)
  queue = [root]
  values = []
  while queue.any?
    node = queue.shift
    values << node.val
    queue << node.left if node.left
    queue << node.right if node.right
  end

  values
end
